### PR TITLE
Filter Enhancement (#933)  - Filter Value Dropdown

### PR DESCRIPTION
**Pull request by the-dcruz**
## Overview

This PR address the issue https://github.com/airbnb/caravel/issues/933, which is to replace the filter TextField with a FreeFormSelect to make it easier for the user to know what types of values are possible for a filter.

If their desired value is not in the list, they can still type in their custom value. (Especially common for regex operation)

The feature must be enabled on the table/datasource.
## Technical Changes
- Added new API endpoint `/filter/<datasource_type>/<datasource_id>/<column>/`
- Added new function `values_for_column` to `models.DruidDatasource` and `models.SqlaTable` (migration script included)
  - `values_for_column` uses the same querystring as `query`
  - Currently limited to **500 values**
- Added new boolean column to `datasources` and `tables` to indicate allowing filter value select
- `add_filter` in `explore.jsx` will now fix the filter IDs upon render instead of by `prepForm` on query submit
## Screenshots

How filter looks:
screen shot 2016-08-15 at 2 07 56 pm

Option on table:
screen shot 2016-08-15 at 2 12 18 pm
## TODO for Future PR
- Entire filter can be redone in React
- Cache filter endpoint

@mistercrunch

### CODE REVIEW

**Comment by the-dcruz** [comment]
This PR is the same as https://github.com/airbnb/caravel/pull/948

Sorry I closed that PR while I was fixing issues after rebasing and my force push didn't allow me to re-open it.

**Comment by kkalyan** [comment]
Great enhancement, thanks @the-dcruz!

Is there a way to select multiple options ? 
Currently comma separated values works as OR. May be switch to Select2 widget ?

**Comment by mistercrunch** [comment]
It needs to be backward compatible with the existing multi-value filters.

**Comment by the-dcruz** [comment]
@kkalyan @mistercrunch 
I currently have this implemented with a freeform select since it is likely that not all possible values for a given column will be displayed (limit=500 currently). My idea was to have this select act as a guide for the user to see a sample of the possible values for the column. The user can then type the values with commas as before. 

Maybe a better implementation I can try would be a freeform multiselect?

**Comment by the-dcruz** [comment]
Updated PR to now use freeform multiselect.

screen shot 2016-09-12 at 1 27 32 pm

**Comment by ShengyaoQian** [comment]
This would be a very helpful enhancement!

**Comment by the-dcruz** [comment]
@mistercrunch I was wondering if there was still a chance of this PR being looked into. I've periodically rebase it but will hold off until I know it will be looked at. Thanks

**Comment by mistercrunch** [comment]
Oh nice! I didn't see your updates since my last comment. I'm a bit overwhelmed with the amount of notifications and communication channels around me.

This looks good, I'll fetch your branch and do some testing on our staging environment today.

**Comment by mistercrunch** [comment]
Quick question before I test, we recently had a PR to fix a bug around filtering on strings that include commas, is that handled on your side?

**Comment by mistercrunch** [comment]
@vera-liu , since you're familiar with this feature and this part of the code, would you mind doing some testing/reviewing here?

**Comment by the-dcruz** [comment]
@mistercrunch Doesn't ring a bell. Could you point me to the PR? I'll try to get that integrated. Thanks for taking a look btw! 

And no worries, I understand it can't be easy managing multiple popular open source projects 😄

**Comment by vera-liu** [comment]
Hi! I was trying to test this on energy_usage example database but somehow the values are not showing up for me. Am I on the right user flow?
1. Toggle 'Enable Filter Select' in Edit mode of energy_usage
2. Go to explore view and click on plus-sign of Filters
3. 
screen shot 2016-09-21 at 1 53 15 pm

**Comment by the-dcruz** [comment]
@vera-liu I use the same time parameters for this query as specified for the slice. Can you show me which time params you selected?

**Comment by vera-liu** [comment]
These ones?
screen shot 2016-09-21 at 2 18 14 pm

**Comment by the-dcruz** [comment]
Yes. Does 7 Days ago actually return any data to display in the chart?

**Comment by vera-liu** [comment]
Yes:
screen shot 2016-09-21 at 2 20 18 pm
But the filter values do not show up in the dropdown:
screen shot 2016-09-21 at 2 21 13 pm

**Comment by the-dcruz** [comment]
@vera-liu I see my issue. Energy usage does not have any time column. 

I guess I'll ignore time in my query if there is no time column. I'll add that fix in a bit. Could you try with something time sensitive like "birth names"?

**Comment by the-dcruz** [comment]
@vera-liu I pushed the fix so it should work for the energy table now. I'll rebase as soon as you think things look good. Thanks!

**Comment by vera-liu** [comment]
It's looking good :). We used to have a bug with comma in filter values: 
https://github.com/airbnb/caravel/issues/1045

Solved in this PR: https://github.com/airbnb/caravel/pull/1084

It would be nice to merge in this PR.

File: caravel/models.py [hunk 1]
```diff
@@ -745,20 +745,21 @@ def values_for_column(self,
...
         qry = qry.limit(limit)
 
+        if granularity:
+            dttm_col = cols[granularity]
+            timestamp = dttm_col.sqla_col.label('timestamp')
+            time_filter = [
+                timestamp >= text(dttm_col.dttm_sql_literal(from_dttm)),
+                timestamp <= text(dttm_col.dttm_sql_literal(to_dttm)),
+            ]
+            qry = qry.where(and_(*time_filter)) if granularity else qry
```
Comment (vera-liu) [inline caravel/models.py]: Why do we need ```if granularity else qry``` here? If this is inside ```if granularity:``` it should always be true right?

Comment (the-dcruz) [inline caravel/models.py]: Good catch. Forgot to remove this.

**Comment by the-dcruz** [comment]
@vera-liu I decided that the most consistent way to integrate your fix is to wrap __all__ the filter choices with the single quotes. 

This seems reasonable to me since this is a dropdown containing column values instead of column names like many of the other dropdowns.

screen shot 2016-09-21 at 5 00 16 pm

**Comment by vera-liu** [comment]
Do you mind rebasing on the master since it has python test for filter values with commas?

**Comment by the-dcruz** [comment]
@vera-liu Sure. Done

**Comment by the-dcruz** [comment]
Looks like another migration file was introduced after rebase, causing the tests to fail. I'll update my migration file and rebase again.

**Comment by the-dcruz** [comment]
Tests look good after rebase @vera-liu

**Comment by vera-liu** [comment]
So Sorry for the delay, I think it's looking good! I'll test it against prod db and then confirm. Thank you for the patience!

**Comment by mistercrunch** [comment]
please rebase so that we can merge this!

**Comment by kkalyan** [comment]
@the-dcruz can't wait to see this in master, please rebase!

**Comment by the-dcruz** [comment]
@kkalyan Sorry about the wait! I had abandoned this PR thinking it was going to be handled waived off until exploreV2.

@mistercrunch I've rebased and retested this.

**Comment by ascott** [comment]
@vera-liu would you make sure this gets added to the filters component in V2?

**Comment by vera-liu** [comment]
Seems like caravel-superset renaming caused conflicts in this PR, I will try to include this feature in explore V2, but feel free to merge into V1 if you would like @the-dcruz,  it's an awesome feature to have!

**Comment by mistercrunch** [comment]
This is done in master

**Pull request closed without merging**